File: tests/e2e/mnist_raycluster_sdk_test.go
/*
Copyright 2023.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package e2e

import (
	"bytes"
	"os/exec"
	"testing"

	. "github.com/onsi/gomega"
	. "github.com/project-codeflare/codeflare-common/support"

	corev1 "k8s.io/api/core/v1"
)

// Creates a Ray cluster, and trains the MNIST dataset using the CodeFlare SDK.
// Asserts successful completion of the training job.
//
// This covers the installation of the CodeFlare SDK, as well as the RBAC required
// for the SDK to successfully perform requests to the cluster, on behalf of the
// impersonated user.
func TestMNISTRayClusterSDK(t *testing.T) {
	test := With(t)
	test.T().Parallel()

	// Create a namespace
	namespace := test.NewTestNamespace()

	// Test configuration
	CreateConfigMap(test, namespace.Name, map[string][]byte{
		// SDK script
		"mnist_raycluster_sdk.py": ReadFile(test, "mnist_raycluster_sdk.py"),
		// pip requirements
		"requirements.txt": ReadFile(test, "mnist_pip_requirements.txt"),
		// MNIST training script
		"mnist.py": ReadFile(test, "mnist.py"),
	})


// TODO: Take first KinD node and redirect pod hostname requests there.
// The below was previously done within the job spec, but here we are no longer using the job spec.

// 	if GetClusterType(test) == KindCluster {
// 	node := GetNodes(test)[0]
// 	hostname := GetClusterHostname(test)
// 	IP := GetNodeInternalIP(test, node)

// 	test.T().Logf("Setting KinD cluster hostname '%s' to node IP '%s' for SDK pod", hostname, IP)
// 	job.Spec.Template.Spec.HostAliases = []corev1.HostAlias{
// 		{
// 			IP:        IP,
// 			Hostnames: []string{hostname},
// 		},
// 	}

// 	// Propagate hostname into Python code as env variable
// 	hostnameEnvVar := corev1.EnvVar{Name: "CLUSTER_HOSTNAME", Value: hostname}
// 	job.Spec.Template.Spec.Containers[0].Env = append(job.Spec.Template.Spec.Containers[0].Env, hostnameEnvVar)
// }
// job, err := test.Client().Core().BatchV1().Jobs(namespace.Name).Create(test.Ctx(), job, metav1.CreateOptions{})
// test.Expect(err).NotTo(HaveOccurred())
// test.T().Logf("Created Job %s/%s successfully", job.Namespace, job.Name)


	// Create a Ray cluster and submit the training job
	cmd := exec.Command("python", "./mnist_raycluster_sdk.py", namespace.Name)

	var stdoutBuf, stderrBuf bytes.Buffer
	cmd.Stdout = &stdoutBuf
	cmd.Stderr = &stderrBuf

	if err := cmd.Run(); err != nil {
		t.Logf("STDOUT: %s", stdoutBuf.String())
		t.Logf("STDERR: %s", stderrBuf.String())
		t.Logf("Failed to run the script: %v", err)
	}


	// Assert the cluster is running
	test.Eventually(func() corev1.ConditionStatus {
		cluster := GetRayCluster(test, namespace.Name, "mnist")
		if cluster == nil {
			// Handle the error appropriately, e.g., by returning an error condition status or logging the error.
			// For simplicity, returning a default status here, but consider proper error handling.
			return corev1.ConditionUnknown
		}
		return corev1.ConditionTrue
	}, TestTimeoutLong).Should(Equal(corev1.ConditionTrue))

	// TODO: Assert the job has been submitted (running) and assert the job has completed successfully.
}


// How it was done previously:
// test.T().Logf("Waiting for Job %s/%s to complete", job.Namespace, job.Name)
// test.Eventually(Job(test, job.Namespace, job.Name), TestTimeoutLong).Should(
// 	Or(
// 		WithTransform(ConditionStatus(batchv1.JobComplete), Equal(corev1.ConditionTrue)),
// 		WithTransform(ConditionStatus(batchv1.JobFailed), Equal(corev1.ConditionTrue)),
// 	))

// // Assert the job has completed successfully
// test.Expect(GetJob(test, job.Namespace, job.Name)).
// 	To(WithTransform(ConditionStatus(batchv1.JobComplete), Equal(corev1.ConditionTrue)))
// }

File: tests/e2e/mnist_raycluster_sdk_test.py
from kubernetes import client, config
import kubernetes.client

import os

from time import sleep

import ray

from codeflare_sdk.cluster.cluster import Cluster, ClusterConfiguration
from codeflare_sdk.job import RayJobClient

import pytest

from support import *

# Creates a Ray cluster, and trains the MNIST dataset using the CodeFlare SDK.
# Asserts creation of AppWrapper, RayCluster, and successful completion of the training job.
# Covers successfull installation of CodeFlare-SDK


@pytest.mark.kind
@pytest.mark.openshift
class TestMNISTRayClusterSDK:
    def setup_method(self):
        initialize_kubernetes_client(self)

    def teardown_method(self):
        delete_namespace(self)

    def test_mnist_ray_cluster_sdk(self):
        create_namespace(self)
        self.run_mnist_raycluster_sdk()

    def run_mnist_raycluster_sdk(self):
        ray_image = get_ray_image()

        cluster = Cluster(
            ClusterConfiguration(
                name="mnist",
                namespace=self.namespace,
                num_workers=1,
                head_cpus="500m",
                head_memory=2,
                min_cpus="500m",
                max_cpus=1,
                min_memory=1,
                max_memory=2,
                num_gpus=0,
                instascale=False,
                image=ray_image,
                write_to_file=True,
                mcad=True,
            )
        )

        cluster.up()
        self.assert_appwrapper_exists()

        cluster.status()

        cluster.wait_ready()
        self.assert_raycluster_exists()

        cluster.status()

        cluster.details()

        auth_token = run_oc_command(["whoami", "--show-token=true"])
        ray_dashboard = cluster.cluster_dashboard_uri()
        header = {"Authorization": f"Bearer {auth_token}"}
        client = RayJobClient(address=ray_dashboard, headers=header, verify=True)

        # Submit the job
        submission_id = client.submit_job(
            entrypoint="python mnist.py",
            runtime_env={
                "working_dir": "./tests/e2e/",
                "pip": "./tests/e2e/mnist_pip_requirements.txt",
            },
        )
        print(f"Submitted job with ID: {submission_id}")
        done = False
        time = 0
        timeout = 900
        while not done:
            status = client.get_job_status(submission_id)
            if status.is_terminal():
                break
            if not done:
                print(status)
                if timeout and time >= timeout:
                    raise TimeoutError(f"job has timed out after waiting {timeout}s")
                sleep(5)
                time += 5

        logs = client.get_job_logs(submission_id)
        print(logs)

        self.assert_job_completion(status)

        client.delete_job(submission_id)
        cluster.down()

    # Assertions
    def assert_appwrapper_exists(self):
        try:
            self.custom_api.get_namespaced_custom_object(
                "workload.codeflare.dev",
                "v1beta1",
                self.namespace,
                "appwrappers",
                "mnist",
            )
            print(
                f"AppWrapper 'mnist' has been created in the namespace: '{self.namespace}'"
            )
            assert True
        except Exception as e:
            print(f"AppWrapper 'mnist' has not been created. Error: {e}")
            assert False

    def assert_raycluster_exists(self):
        try:
            self.custom_api.get_namespaced_custom_object(
                "ray.io", "v1", self.namespace, "rayclusters", "mnist"
            )
            print(
                f"RayCluster 'mnist' created successfully in the namespace: '{self.namespace}'"
            )
            assert True
        except Exception as e:
            print(f"RayCluster 'mnist' has not been created. Error: {e}")
            assert False

    def assert_job_completion(self, status):
        if status == "SUCCEEDED":
            print(f"Job has completed: '{status}'")
            assert True
        else:
            print(f"Job has completed: '{status}'")
            assert False
